controllers/cloudstackmachine_controller.go

```diff
@@ -268,6 +269,17 @@ func (r *CloudStackMachineReconciliationRunner) GetOrCreateMachineStateChecker()
 }
 
 func (r *CloudStackMachineReconciliationRunner) ReconcileDelete() (retRes ctrl.Result, reterr error) {
+	if r.ReconciliationSubject.Spec.InstanceID == nil {
+		// InstanceID is not set until deploying VM finishes which can take minutes, and CloudStack Machine can be deleted before VM deployment complete.
+		// ResolveVMInstanceDetails can get InstanceID by CS machine name
+		err := r.CSClient.ResolveVMInstanceDetails(r.ReconciliationSubject)
+		if err != nil {
```

Comment:
what if the VM isn't present at all in ACS, if it was never created? We may want to exit then

Reply:
This is a tough question - whether we exit or not.
For example, ee got an 'no match found' error
case 1:
This could be a race condition, the VM may in the process of deployment, but not done. The next retry may found it.
case 2:
This VM could be manually deleted through other ways, like CloudStack UI. And this `no match found` could be permanent. Controller enters into rate limited retry loop(up to once every around 16 mins)

Here we follow the best practice:

- set error in status/reason

- Keep retry
Eventually, admin can remove the finalizer, and CloudStackMachine will be GCed by kubernetes runtime. 

A similar situation in Kubernetes is pod having an invalid image configuration, and keeps getting crashloopbackoff until admin fix it.

Reply:
The pod image configuration is a different situation - it's not a race condition. It will only happen if a cluster is misconfigured by the user, and would only happen once. The operational implications are different. I think we should be more optimistic/automated with this race condition problem and exit on 'no match found' error instead of naively retrying. In a slow environment, I believe we're very likely to have CAPI machines created/deleted without a corresponding ACS VM. I do not think we should require a manual intervention every time this happens, it will be way too much overhead to be sustainable, with create/upgrade/delete operations failing nondeterministically now.

Reply:
For race-condition situation, we don't have a problem here. The next retry will get the instanceID and cloudstack VM will be deleted and its associated cloudstackmachine finalizer will be removed -- everything is fine here.

Here we are trying to address cloudstackmachine object is deleted while leaving cloudstack VM not being destroyed. I don't know a situation that a cloudstackmachine created without having a cloudstack VM -- if this happens, the controller will keep retrying to create this VM.

Do you have any reason for `I believe we're very likely to have CAPI machines created/deleted without a corresponding ACS VM` ?

Reply:
I think the challenge is to distinguish above 2 cases I described, even we exit (with/without error), I assume it is case 2 and it is a risky assumption.

Reply:
My thinking is this - suppose a client creates 1000 CAPI machines all at once. ACS will put those requests into a queue right? Suppose before the MHC timeout, it successfully creates 50 VM’s. Then MHC goes in and tries to delete all the timed out CAPI machines. Doesn’t that leave us in a situation where 950 CAPI machines with no associated VM need to be manually deleted by an admin? My concern is that’s a ton of overhead

Reply:
In this condition, I think no manual overhead.

950 Cloudstackmachine object will be created in etcd, and CAPC controller will try to reconcile and deploy 950 VMs in cloudstack - it takes time, but before it finishes, deleting (caused by MHC) those cloudstackmachine will return error and retry.

DeployVM can fail with an InstanceID which can be later retrieved by ResolveVMInstanceDetails.

Reply:
What's the termination condition in this scenario? When does the CAPC VM creation workflow terminate? When does this deletion workflow terminate? I'm struggling to see a clean outcome for all the CloudStackMachines being cleanly deleted when failures occur during both VM creation and deletion. Can you explain again? or I'd be happy to discuss over slack

Reply:
During cloudstackmachine creation: the CAPI machine will be deleted right away (no MHC involve) when VM deployment failed (the instancestate will be error), in this case, we still have a VM instanceID. So the CAPI machine and VM (in error state) will be cleaned up.

During cloudstackmachine deletion: CAPC controller will try to destroy using instanceID, if destroying failed, it will keep retrying. if failed with error message `not found`, it considered success.
CAPC controller will keep retrying to get instanceID when instanceID is missing.